### ISSUE TITLE
Support for inserting uploaded file content in db as bytes for streaming upload/download

### ISSUE THREAD

**Issue by djyotta** [sqlpage/SQLPage#389]
**_What are you building with SQLPage ?_**

A tool to share files.

**_What is your problem ? A description of the problem, not the solution you are proposing._**

Upload: I can't do streaming upload into the db because the data_uri is manifested as a string in memory first.

Download: I use an external service to provide a streaming download of the data. But this is not as ideal as it could be because I have to decode the data_uri from the db first (buffer in memory before download starts). Theoretically, one could decode a data_uri as a stream, but ideally I could just stream the bytes of the blob as is.

**_What are you currently doing ? Since your solution is not implemented in SQLPage currently, what are you doing instead ?_**

I'm storing uploaded files as data uri in db and using an external service to provide the download.

**Describe the solution you'd like**

I'd like to store the bytes alone as blob in db. Ideally this could be inserted into db as a streaming upload to avoid buffering entire file in memory.

~~SQLite buffers blobs in memory anyway, but other dbs may not.~~ Even sqlite supports streaming blobs https://www.sqlite.org/c3ref/blob_open.html

Actually, streaming download could be a great featur for SQLPage (a new component maybe like the JSON component...?).

**Comment by lovasoa** [comment]
Hi! Streaming blobs in and out of databases is not supported in SQL. Databases have their own interfaces, but that would require some non-sql syntax. Storing large files in databases is usually not recommended anyway, so I'm not sure there would be enough use of such a feature to justify its development...

**Comment by djyotta** [comment]
Yeah, I was prior to SQLPage storing everything on disk. I rolled my own content addressable storage.
But most these files are tiny so the db makes more sense.

Regarding large files, in db - I think it really depends on the db whether it's good or bad. Personally, I'd rather store files in a db than on disk. In my case, I have a SBC so storage is limited. With a db connection, the storage could be anywhere. Yes, I could also use a NAS, but a db seems way more portable.

My vision for such a SQLPage feature is like this:
```sql
-- index.sql

-- download (probably has to be first component like for JSON component
SET ":key" = $download;
SELECT 'streaming-download' AS component, (SELECT sqlpage.streaming_download('large_files', :key)) AS content
WHERE $download IS NOT NULL;

-- upload form
SELECT 'form' AS component;
SELECT 'file' AS type, 'upload' AS name, 'Upload' AS label;


-- upload
SET ":key" = sqlpage.uploaded_file_name();
INSERT INTO large_files (key, content)
VALUES (:key, sqlpage.empty_blob());
-- stream the bytes into the zero-sized blob
SELECT sqlpage.stream_uploaded_file_to_db(:key, sqlpage.uploaded_file_path('upload'))
WHERE :upload IS NOT NULL;

-- display a button to retrieve the just uploaded content
SELECT 'button' AS component, '/index.sql?download='||:key AS link;
```

So still SQL, but of course those functions need to be implemented - and likely need to be specifically implemented per db...

But yeah, I get this is likely not within scope of SQLPage target feature set.

**Comment by lovasoa** [comment]
Hey ! I think we can close this now that v0.37 has been released :) https://github.com/sqlpage/SQLPage/releases/tag/v0.37.0

**Comment by djyotta** [comment]
@lovasoa I see we can read BLOBs as data uri, and (maybe?) the blob is streamed from the download component. I wonder can we insert data uri into BLOB type column too?

I'm revisiting this as it looks like I could use this for downloads, but maybe not for uploads. Just want to avoid false assumptions before I get started...

I intend to start with a dummy app to handle file upload/download to BLOB type column and see how far I get. Having a good idea what to expect to work will help me avoid trying to do something that could never work to start with.

_yes, I did RTFM but I could probably RTFM harder. The functionality wasn't immediately obvious from the docs I've read so far_

**Comment by djyotta** [comment]
OK, I think I have answered my own question.
This is a working example of upload file _direct_ (*) to db:
```sql
SET ":file_name" = CASE :file WHEN NULL THEN NULL ELSE sqlpage.uploaded_file_name('file') END;                                                                    SET ":file_path" = CASE :file WHEN NULL THEN NULL ELSE sqlpage.uploaded_file_path('file') END; 
                                                                 
-- save
INSERT INTO myfiles (name, content)
SELECT
    :file_name,
    -- readfile() is provided by fileio sqlite extension
    readfile(:file_path)
WHERE :file_path IS NOT NULL;

SELECT 'redirect' AS component, '/myfiles.sql' AS link
WHERE :file_path IS NOT NULL;

-- upload form
SELECT 'form' AS component,
       'Upload File' AS title
WHERE :file_path IS NULL;
SELECT 'file' AS type,
       'file' AS name,
       'Choose a file' AS label
WHERE :file_path IS NULL;
```
\* actually, there is a temp file created for the upload - not sure if it's a stream or not

So upload of BLOB can be done without external tools with sqlpage and sqlite backend with fileio extension enabled.

I did have to compile the fileio.so from source on ARMv7 though.
